File: src/main/java/io/example/documentapproval/repository/UserRepository.java
package io.example.documentapproval.repository;

import io.example.documentapproval.domain.User;
import org.springframework.data.jpa.repository.EntityGraph;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;
import java.util.Set;

@Repository
public interface UserRepository extends JpaRepository<User, Long> {

    List<User> findByIdIn(Set<Long> userIds);

    @EntityGraph(attributePaths = {"authoritySet"})
    Optional<User> findByEmail(String email);

    <T> List<T> findAllProjectedBy(Class<T> type);

}

File: src/main/kotlin/com/example/document/infra/repository/UserRepository.kt
package com.example.document.infra.repository

import com.example.document.domain.User
import com.example.document.infra.config.R2dbcRepository
import org.springframework.data.repository.kotlin.CoroutineSortingRepository

@R2dbcRepository
interface UserRepository : CoroutineSortingRepository<User, Long> {

    suspend fun findById(id: Long): User?
}
